server/src/main/java/org/opensearch/index/store/remote/utils/cache/LinkedDeque.java

```diff
@@ -0,0 +1,432 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.store.remote.utils.cache;
+
+import java.util.AbstractCollection;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * Linked list implementation of the {@link Deque} interface where the link
+ * pointers are tightly integrated with the element. Linked deques have no
+ * capacity restrictions; they grow as necessary to support usage. They are not
+ * thread-safe; in the absence of external synchronization, they do not support
+ * concurrent access by multiple threads. Null elements are prohibited.
+ * <p>
+ * Most LinkedDeque operations run in constant time by assuming that
+ * the {@link Linked} parameter is associated with the deque instance. Any usage
+ * that violates this assumption will result in non-deterministic behavior.
+ * <p>
+ * The iterators returned by this class are <em>not</em> <i>fail-fast</i>: If
+ * the deque is modified at any time after the iterator is created, the iterator
+ * will be in an unknown state. Thus, in the face of concurrent modification,
+ * the iterator risks arbitrary, non-deterministic behavior at an undetermined
+ * time in the future.
+ *
+ * @param <E> the type of elements held in this collection
+ *
+ * @opensearch.internal
+ */
+public final class LinkedDeque<E extends Linked<E>> extends AbstractCollection<E> implements Deque<E> {
```

Comment:
Is it  copy of [1]? Why?

> [Caffeine](https://github.com/ben-manes/caffeine) is the Java 8 successor to ConcurrentLinkedHashMap and Guava's cache. Projects should prefer Caffeine and migrate when requiring JDK8 or higher. The previous caching projects are supported in maintenance mode.


[1] https://github.com/ben-manes/concurrentlinkedhashmap/blob/master/src/main/java/com/googlecode/concurrentlinkedhashmap/LinkedDeque.java

Reply:
most probably it was copied from Guava. Look at https://github.com/opensearch-project/OpenSearch/pull/5641#discussion_r1093867050 and https://github.com/opensearch-project/OpenSearch/issues/4964#issuecomment-1421689586 for more context.

Reply:
and here https://github.com/opensearch-project/OpenSearch/pull/5641#discussion_r1099493022